### PR TITLE
Improve JAX-WS models

### PR DESCRIPTION
**Pull request by smowton**
This includes @lcartey's definition of JAX-WS sinks, plus models for the various wrapper types it uses to represent user-controlled data.

Contributes to https://github.com/github/codeql-java-team/issues/37

Other JAX tasks to do in separate PRs:

* ~Note open-redirect sinks in ResponseBuilder~ (now done here)
* Add tests (@owen-mc is currently working on this)
* ~Determine whether the framework promises to sanitise certain e.g. URI parts -- for example, is it possible to use UriBuilder.scheme(...) to inject something dangerous, or does the contract promise we'll check you're writing one of http, https, file ... rather than `<script>...</script>`~ (assuming these are dangerous for now, as they're allowed to contain template parameters in particular, which suggests quite a broad acceptable charset)

### CODE REVIEW

File: java/ql/src/semmle/code/java/frameworks/JaxWS.qll [hunk 1]
```diff
@@ -1,4 +1,10 @@
 import java
+private import semmle.code.java.dataflow.ExternalFlow
+
+string getAJaxWsPackage() { result in ["javax.ws.rs", "jakarta.ws.rs"] }
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/frameworks/JaxWS.qll:4]: This is JAX-<del>W</del><ins>R</ins>S

File: java/ql/src/semmle/code/java/security/UrlRedirect.qll [hunk 2]
```diff
@@ -35,3 +36,17 @@ private class ApacheUrlRedirectSink extends UrlRedirectSink {
     )
   }
 }
+
+/** A URL redirection sink from JAX-WS */
+private class JaxWsUrlRedirectSink extends UrlRedirectSink {
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/security/UrlRedirect.qll:41]: This is JAX-<del>W</del><ins>R</ins>S

File: java/change-notes/2021-03-22-jax-ws-improvements.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+lgtm,codescanning
+* Added support for detecting XSS via JAX-WS sinks, and propagating tainted data via various container types (e.g. Form, Cookie, MultivaluedMap).
```
Comment (Marcono1234) [inline java/change-notes/2021-03-22-jax-ws-improvements.md:2]: This is JAX-<del>W</del><ins>R</ins>S

**Review by Marcono1234 — COMMENTED** [review]
There seems to be some confusion between [JAX-RS](https://en.wikipedia.org/wiki/Jakarta_RESTful_Web_Services) and [JAX-WS](https://en.wikipedia.org/wiki/Jakarta_XML_Web_Services), see also  #5559.
(Unless I am overlooking something)

## JAX-RS
"Jakarta RESTful Web Services" (formerly "Java API for RESTful Web Services")
[Packages](https://jakarta.ee/specifications/restful-ws/3.0/apidocs/):
- `jakarta.ws.rs`
- `jakarta.ws.rs.*`

## JAX-WS
"Jakarta XML Web Services" (formerly "Java API for XML Web Services")

[Packages](https://jakarta.ee/specifications/xml-web-services/3.0/apidocs/jakarta.xml.ws/module-summary.html):
- `jakarta.xml.ws`
- `jakarta.xml.ws.*`

Additionally there is [Jakarta Web Services Metadata API](https://jakarta.ee/specifications/web-services-metadata/3.0/apidocs/jakarta.jws/module-summary.html):
- `jakarta.jws`
- `jakarta.jws.soap`

---

Maybe it would be good to create a separate `JaxRS.qll` to avoid future confusion?

**Comment by smowton** [comment]
Thanks @Marcono1234, fixed

**Comment by Marcono1234** [comment]
Thanks!

>Maybe it would be good to create a separate `JaxRS.qll` to avoid future confusion?

And what is your opinion on this? Maybe not necessary as part of this pull request, if you want I can also create a new GitHub issue for tracking this.

**Comment by smowton** [comment]
Perhaps. Potentially breaks customer users, and only `JaxWsEndpoint` would move.

**Comment by smowton** [comment]
I added a doc comment to the file mentioning and distinguishing the two standards

**Comment by owen-mc** [comment]
Superseded by https://github.com/github/codeql/pull/6034

**Comment by owen-mc** [comment]
Superseded by https://github.com/github/codeql/pull/6034

**Pull request closed without merging**